tests/unit/cluster/slot-migration.tcl

```diff
@@ -0,0 +1,355 @@
+        assert_equal {OK} [R 3 CLUSTER REPLICATE $R0_id]
+        wait_for_role 3 slave
+        # Validate that R3 now sees slot 609 open
+        assert_equal [get_open_slots 3] "\[609->-$R1_id\]"
+    }
+
+    test "New replica inherits importing slot" {
+        # Reset R4 to turn it into an empty node
+        assert_equal [get_open_slots 4] "\[609-<-$R0_id\]"
+        assert_equal {OK} [R 4 CLUSTER RESET]
+        assert_not_equal [get_open_slots 4] "\[609-<-$R0_id\]"
+        # Add R4 back as a replica of R1
+        assert_equal {OK} [R 4 CLUSTER MEET [srv -1 "host"] [srv -1 "port"]]
+        wait_for_role 1 master
+        assert_equal {OK} [R 4 CLUSTER REPLICATE $R1_id]
+        wait_for_role 4 slave
+        # Validate that R4 now sees slot 609 open
+        assert_equal [get_open_slots 4] "\[609-<-$R0_id\]"
+    }
+}
+
+proc create_empty_shard {p r} {
+    set node_timeout [lindex [R 0 CONFIG GET cluster-node-timeout] 1]
+    assert_equal {OK} [R $p CLUSTER RESET]
+    assert_equal {OK} [R $r CLUSTER RESET]
+    assert_equal {OK} [R $p CLUSTER MEET [srv 0 "host"] [srv 0 "port"]]
+    assert_equal {OK} [R $r CLUSTER MEET [srv 0 "host"] [srv 0 "port"]]
+    wait_for_role $p master
+    assert_equal {OK} [R $r CLUSTER REPLICATE [R $p CLUSTER MYID]]
+    wait_for_role $r slave
+    wait_for_role $p master
+}
+
+start_cluster 3 5 {tags {external:skip cluster} overrides {cluster-allow-replica-migration no cluster-node-timeout 1000} } {
+
+    set node_timeout [lindex [R 0 CONFIG GET cluster-node-timeout] 1]
+    set R0_id [R 0 CLUSTER MYID]
+    set R1_id [R 1 CLUSTER MYID]
+    set R2_id [R 2 CLUSTER MYID]
+    set R3_id [R 3 CLUSTER MYID]
+    set R4_id [R 4 CLUSTER MYID]
+    set R5_id [R 5 CLUSTER MYID]
+
+    create_empty_shard 6 7
+    set R6_id [R 6 CLUSTER MYID]
+    set R7_id [R 7 CLUSTER MYID]
+
+    test "Empty-shard migration replicates slot importing states" {
+        # Validate initial states
+        assert_not_equal [get_open_slots 0] "\[609->-$R6_id\]"
+        assert_not_equal [get_open_slots 6] "\[609-<-$R0_id\]"
+        assert_not_equal [get_open_slots 3] "\[609->-$R6_id\]"
+        assert_not_equal [get_open_slots 7] "\[609-<-$R0_id\]"
+        # Kick off the migration of slot 609 from R0 to R6
+        assert_equal {OK} [R 0 CLUSTER SETSLOT 609 MIGRATING $R6_id]
+        assert_equal {OK} [R 6 CLUSTER SETSLOT 609 IMPORTING $R0_id]
+        # Validate that R0 is migrating slot 609 to R6
+        assert_equal [get_open_slots 0] "\[609->-$R6_id\]"
+        # Validate that R6 is importing slot 609 from R0 
+        assert_equal [get_open_slots 6] "\[609-<-$R0_id\]"
+        # Validate final states
+        wait_for_slot_state 0 "\[609->-$R6_id\]"
+        wait_for_slot_state 6 "\[609-<-$R0_id\]"
+        wait_for_slot_state 3 "\[609->-$R6_id\]"
+        wait_for_slot_state 7 "\[609-<-$R0_id\]"
+    }
+
+    test "Empty-shard migration target is auto-updated after faiover in target shard" {
```

Comment:
How will `valkey-cli` behave if it's in the middle of transferring keys between the nodes when the failover happens? WIll it seamlessly continue and finish the slot after failover?

Reply:
I'm starting to think that we could let a cron job could finish the (non-atomic) slot migration, i.e. transfer keys. Without `valkey-cli` involved, it might be easier to get a self-healing cluster.

Reply:
All good points and if you remember our discussion back in the Redis days these were all on my plan

1. Introduce "cluster migrate slot"
2. Have engine drive the migration in serverCron
3. Auto resume on failures

I also believe the same interface can be reused for atomic slot migration without causing breaks.

Let's file separate issues?

Reply:
Yeah, incremental perfection. .)

For now, I just would like to know how this PR affects the documented `valkey-cli`-based flow. Can it resume a slot-migration? It quite possibly can. It'd just be good to test it.

Reply:
>For now, I just would like to know how this PR affects the documented valkey-cli-based flow. Can it resume a slot-migration? It quite possibly can. It'd just be good to test it.

From a quick look at `clusterManagerMigrateKeysInSlot`, it does not look like that valkey-cli would be able to resume on failovers. `MIGRATE` is a write command that can only be executed on the primary but I haven't seen valkey-cli rediscovering the topology in `clusterManagerMigrateKeysInSlot`?

Reply:
No, `--cluster fix` maybe. If not, I think we should probably add something there to discover and resume an aborted migration.